### ISSUE TITLE
Constructor metamodels revisited

### ISSUE THREAD

**Issue by tombentley** [ceylon/ceylon.language#718]
In https://github.com/ceylon/ceylon-compiler/issues/2205 it became apparent that we haven't yet dealt with the metamodel for value constructors. To summarize what I'm thinking, given a class like this:

    class C {
        shared new () {}
        shared new ctor() {}
        shared new val {}
    }

We would have the following possibilities for meta/declaration literals

Literal | Meta type
--------|------------
`` `C` ``  | `Class<C,[]>` // default constructor
`` `C.ctor` ``  | `Function<C,[]>` // ctor
`` `C.val` ``  | `Value<C>`
`` `class C` `` | `ClassWithConstructorsDeclaration`
`` `new Class` `` | `ConstructorDeclaration` // the default ctor
`` `new Class.ctor` `` | `ConstructorDeclaration`
`` `new Class.val` `` | `ConstructorDeclaration`
`` `function Class.ctor` `` | `FunctionDeclaration`
`` `value Class.val` `` | `ValueDeclaration`

For a member class `` `C.MC.ctor` `` would be a `Method<C,MC,arguments>` and `` `C.MC.val` `` would be an `Attribute<C,MC>`.

Thus it seems to me that we don't actually need the `Constructor` or `MemberConstructor` models any more. 

`ConstructorDeclaration.apply()` would return a `Function|Value` and `ConstructorDeclaration.memberApply()` would return a `Method|Attribute`

Does that seem a reasonable to everyone?

**Comment by jvasileff** [comment]
How would `FunctionModel.declaration` work for constructors? Currently it returns `FunctionDeclaration`. (Same idea for `ValueModel`).

**Comment by tombentley** [comment]
I suppose we could make those return the same `FunctionDeclaration` (`ValueDeclaration`) that I'm proposing we have for `` `function C.ctor` `` and `` `value C.val` ``. Making `FunctionModel.declaration` a union type (e.g. `FunctionDeclaration|ConstructorDeclaration`) would be unhelpful in the much more usual case of dealing with a "normal" function/method/value/attribute, so I don't think we should do that.

If we have two representations for the declarations of constructors, it raises the question of what the `ConstructorDeclaration` looks like and what it is really useful for. I think the main use case is for constraining annotations to constructors, so maybe `ConstructorDeclaration` looks something like this:

```ceylon
shared sealed interface ConstructorDeclaration {
    "The constructors name."
    shared formal String name;

    "The declaration of the class this constructors it within."
    shared formal ClassDeclaration container;

    "The representation of this constructor as a function or value."
    // nasty name
    shared formal ValueDeclaration|FunctionDeclaration functionOrValueDeclaration;
    
    "The declaration of the constructor or class that appears 
     in this constructors `extends` clause"
    shared formal ConstructorDeclaration|ClassDeclaration extendedDeclaration;
}
```

We *could* have a `FunctionalConstructorDeclaration` (a subtype of `FunctionDeclaration`) and `ValueConstructorDeclaration` (a subtype of `ValueDeclaration`), but that would mean they were subtypes of `NestableDeclaration`, which would be wrong.

@FroMage and/or @gavinking must have an opinion about this.

**Comment by gavinking** [comment]
For consistency it seems like we should probably have `CallableConstructorDeclaration` and `ValueConstructorDeclaration`. I'm not sure what their supertypes should be.

**Comment by tombentley** [comment]
Reflecting on this yesterday I did feel like having two kinds of declaration for each constructor would eventually be seen as a mistake. Thus really `CallableConstructorDeclaration` ought to be a subtype of `FunctionDeclaration` and `ValueConstructorDeclaration` a subtype of `ValueDeclaration` and we should try to fix the rest of the API to make this possible. Those fixes include:

* `FunctionDeclaration` indirectly inherits from `NestableDeclaration`, but a constructor can only appear in a class. A `NestableDeclaration` is supposed to be 

    > A declaration which can be contained in a Package or in another NestableDeclaration.

    It's not clear how much of a problem this really is.

* `FunctionDeclaration` inherits from `GenericDeclaration`, but a constructor is not generic.

But `GenericDeclaration` has its own problems: It seems odd that it doesn't declare `apply()`. Likewise it seems odd that `FunctionalDeclaration` doesn't declare `invoke()`. Instead `ClassDeclaration` declares `instantiate(Type[], Anything[])` and `FunctionDeclaration` declares `invoke(Type[], Anything[])` (note how the genericness manifests in the `Type[]` parameter).

**Comment by tombentley** [comment]
I seem to have hit yet another problem with the class metamodel.
`CallableConstructorDeclaration.apply()` should return a `Function`, because `CallableConstructorDeclaration` satisfies `FunctionalDeclaration`. Which is all grand except in the case of the default constructor. For that we really want `apply()` to return the `Class`, because we want `` `ClassWithParameters` `` and `` `ClassWithADefaultConstructor` `` to both be a `Class` (so you can retrospectively constructorize a class).
Strictly speaking I can keep `` `ClassWithADefaultConstructor` `` being a Class, but then `` `new ClassWithADefaultConstructor`.apply() `` is a `Function` rather than the same `Class`, which is at least weird.

So I suppose the solution to this is to not represent the default constructor as a `CallableConstructorDeclaration`, but as a `ClassWithConstructorsDeclaration`. So `` `new ClassWithAConstructor` `` is a `ClassWithConstructorsDeclaration`. But it's a bit weird seeing ``ClassWithConstructorsDeclaration foo = `new ClassWithAConstructor` `` because we're very used to there being a fixed mapping from the  declaration literals to a single corresponding declaration meta type. Maybe we should forbid `` `new ClassWithAConstructor` ``.

**Comment by tombentley** [comment]
@gavinking pointed out that this doesn't really work from the pov of annotations:

* It's possible to constrain an annotation to the non-default callable constructors, which is not a requirement
* Worse it means you can't really obtain the constructors annotations, only the classes annotations.

So we at least need a separate object to represent the constructor so we can obtain the annotations on the default constructor at runtime.

**Comment by tombentley** [comment]
But in order to allow refactoring a class with initializer parameters into a class with a default constructor without breaking client metamodel code we really need to make the default constructor look like a `ClassDeclaration` and `Class`. Or do people think not breaking client metamodel code is going to be mostly impossible anyway?

**Comment by gavinking** [comment]
@tombentley we already have a split between `ClassWithInitializer` and `ClassWithConstructors`. I feel like that should be enough to solve your problem.

**Comment by tombentley** [comment]
I don't see how that really helps much, to be honest.

**Comment by tombentley** [comment]
I've been putting this off because I still don't really have a clear idea of how to model constructors in the metamodel. Or rather, what is the least worst way of doing it. For example, right now `ClassModel` satisfies `Functional` which means it has `parameterTypes`, but a class with constructors but no default constructor doesn't have parameters in any meaningful way, so we can't really keep that. Likewise `ClassDeclaration` satisfies `FunctionalDeclaration` so has `parameterDeclarations`. 

I don't really want to embark on refactoring until I have a clear idea of where I'm headed, because this stuff takes days to do (and more so the more we change from the existing model). But to have that clear idea I need some kind of consensus about the best way to model this stuff. I know @gavinking hates the idea of `Class` not being `Callable`, for example, but the fact is that we can have non-callable classes now, and @FroMage seemed to think that non-`Callable` classes was the way to go.

**Comment by tombentley** [comment]
I've pushed changes to the `ctor-model` branches of ceylon-spec, ceylon-model, ceylon-compiler and ceylon.language which define an API for the metamodel of constructors. It's taken me a long time to get an API which works and it's still not entirely complete, but I think it's complete enough to let @chochos have a go at implementing it in JS. 

To get here I tried lots of simpler things (for example having CallableConstructorDeclaration inherit FunctionDeclaration -- that doesn't let you constrain an annotation to a function-but-not-a-constructor) which all didn't work for one reason or another.
